src/wp-includes/l10n.php

```diff
@@ -1988,3 +1988,118 @@ function wp_get_word_count_type() {
 
 	return $wp_locale->get_word_count_type();
 }
+
+/**
+ * Parses the `Accept-Language` header to get a list of locales.
+ *
+ * The locales are returned in the format WordPress expects,
+ * so "fr-CH" becomes "fr_CH" and "fr" becomes "fr_FR".
+ *
+ * @link https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language
+ *
+ * @since 6.7.0
+ * @access private
+ *
+ * @global wpdb $wpdb WordPress database abstraction object.
+ *
+ * @return string[] Locales list.
+ */
+function get_locales_from_accept_language_header() {
+	global $wpdb;
+
+	$locales = array();
+
+	if ( ! empty( $_SERVER['HTTP_ACCEPT_LANGUAGE'] ) && is_string( $_SERVER['HTTP_ACCEPT_LANGUAGE'] ) ) {
+		$matches = array();
+		// Parses a header value such as "fr-CH, fr;q=0.9, en;q=0.8, de;q=0.7, *;q=0.5".
+		preg_match_all( '((?P<code>[a-z-_A-Z]+|\*)([;q=]+?(?P<prio>1|0\.\d))?)', $_SERVER['HTTP_ACCEPT_LANGUAGE'], $matches );
+
+		if ( empty( $matches['code'] ) ) {
+			return $locales;
+		}
+
+		$codes = $matches['code'];
+
+		// An empty priority defaults to 1.
+		$prios = array_map(
+			static function ( $value ) {
+				if ( '' === $value ) {
+					return 1.0;
+				}
+
+				return (float) $value;
+			},
+			$matches['prio']
+		);
+
+		// Sort codes by priority.
+		usort(
+			$codes,
+			static function ( $a, $b ) use ( $codes, $prios ) {
+				$index_a = array_search( $a, $codes, true );
+				$index_b = array_search( $b, $codes, true );
+
+				return $prios[ $index_b ] <=> $prios[ $index_a ];
+			}
+		);
+
+		$translations = array();
+
+		/*
+		 * Get list of available translations without potentially deleting an expired transient and causing an HTTP request.
+		 * Only works if either the object cache or the database are already available.
+		 */
+		if ( wp_using_ext_object_cache() || wp_installing() ) {
+			wp_start_object_cache();
+			$translations = wp_cache_get( 'available_translations', 'site-transient' );
+		} elseif ( isset( $wpdb ) ) {
+			$translations = get_site_option( '_site_transient_available_translations' );
+		}
```

Comment:
Not exactly pretty, but needed to ensure `wp_cache_get` works, or that there is no database query being attempted if the db is not set up.